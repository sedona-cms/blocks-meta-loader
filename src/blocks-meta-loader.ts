import * as fs from 'fs'
import * as path from 'path'

export class BlocksMetaLoader {
  async getMetaFromFile(filePath: string | string[]): Promise<boolean> {
    if (typeof filePath === 'string') {
      return await this.parseSingleFile(filePath)
    } else if (Array.isArray(filePath)) {
      const promises = filePath.map(item => this.parseSingleFile(item))
      const result = await Promise.all(promises)
      console.log(result)
    }

    return true
  }

  getMetaFromDirectory(directoryPath: string) {}

  private async parseSingleFile(filePath: string): Promise<any> {
    if (!fs.existsSync(filePath)) {
      throw new Error(`Block file does not exists in ${filePath}`)
    }
    return Promise.resolve('ok')
  }
}
