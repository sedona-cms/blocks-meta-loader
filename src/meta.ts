export type BlockMeta = {
  name?: string
  title?: string
  description?: string
  icon?: string // default is 'extension'
  group: string // default is 'general'
  path?: string // path to component file, use for import
  props: {
    [key: string]: BlockProp
  }
}

export type BlockProp = {
  title?: string
  type?: 'string' | 'boolean' | 'number' | 'array' | 'object' | 'date' | string
  editor?: 'text' | 'textarea' | 'number' | 'checkbox' | 'date' | string
  required?: boolean
  default?: unknown
  options?: unknown
}
