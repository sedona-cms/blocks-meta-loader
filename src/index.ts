import { BlocksMetaLoader } from './blocks-meta-loader'

export { BlocksMetaLoader }
