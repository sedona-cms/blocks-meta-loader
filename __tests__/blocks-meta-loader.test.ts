import * as path from 'path'
import { BlocksMetaLoader } from '../src'

const blocksMetaLoader = new BlocksMetaLoader()

test('load block from single file', async () => {
  const blockFilePath = path.resolve(__dirname, 'stubs/text-block.vue')
  const blockMeta = await blocksMetaLoader.getMetaFromFile(blockFilePath)

  expect(blockMeta).toMatchSnapshot()
})

test('load block from single js file', async () => {
  const blockFilePath = path.resolve(__dirname, 'stubs/feature-block.js')
  const blockMeta = await blocksMetaLoader.getMetaFromFile(blockFilePath)

  expect(blockMeta).toMatchSnapshot()
})

test('load block from not exists single file', () => {
  const blockFilePath = path.resolve(__dirname, 'stubs/wrong-file-path.vue')

  expect(blocksMetaLoader.getMetaFromFile(blockFilePath)).rejects.toThrow()
})

test('load block with bad type argument', () => {
  // @ts-ignore
  expect(blocksMetaLoader.getMetaFromFile(123)).rejects.toThrow()
})

test('load block from list files', async () => {
  const blockMeta = await blocksMetaLoader.getMetaFromFile([
    path.resolve(__dirname, 'stubs/text-block.vue'),
    path.resolve(__dirname, 'stubs/feature-block.js'),
  ])

  expect(blockMeta).toMatchSnapshot()
})

test('load blocks from directory', async () => {
  const blocksPath = path.resolve(__dirname, 'stubs')
  const blockMeta = await blocksMetaLoader.getMetaFromDirectory(blocksPath)

  expect(blockMeta).toMatchSnapshot()
})

test('load blocks from wrong directory', () => {
  const blocksPath = path.resolve(__dirname, 'wrong-path')

  expect(blocksMetaLoader.getMetaFromDirectory(blocksPath)).rejects.toThrow()
})

test('load block from vue file without a meta information provided', async () => {
  const blocksPath = path.resolve(__dirname, 'stubs/hero-block.vue')
  const meta = await blocksMetaLoader.getMetaFromFile(blocksPath)

  expect(meta).toHaveProperty('name', 'HeroBlock')
  expect(meta).toHaveProperty('title', 'Hero Block')
  expect(meta).toHaveProperty('group', 'general')
  expect(meta).toHaveProperty('props', {})
  expect(meta).toHaveProperty('path', 'hero-block')
})
