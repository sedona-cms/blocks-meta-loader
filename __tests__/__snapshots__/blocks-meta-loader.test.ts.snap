// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load block from list files 1`] = `
Array [
  Object {
    "group": "general",
    "icon": "extension",
    "name": "TextBlock",
    "path": "text-block",
    "props": Object {
      "fontSize": Object {
        "default": 12,
        "editor": "number",
        "options": Object {},
        "required": false,
        "title": "Font Size",
        "type": "number",
      },
      "isDark": Object {
        "default": false,
        "editor": "checkbox",
        "options": Object {},
        "required": false,
        "title": "Is Dark",
        "type": "boolean",
      },
      "items": Object {
        "default": Array [],
        "editor": "textarea",
        "options": Object {},
        "required": false,
        "title": "Items",
        "type": "array",
      },
      "text": Object {
        "default": "default text",
        "editor": "text",
        "options": Object {},
        "required": false,
        "title": "Text",
        "type": "string",
      },
    },
    "title": "Text",
  },
  Object {
    "group": "general",
    "name": "FeatureBlock",
    "path": "feature-block",
    "props": Object {
      "color": Object {
        "default": "",
        "editor": "text",
        "required": true,
        "title": "Color",
        "type": "string",
      },
      "hasTitle": Object {
        "default": false,
        "editor": "checkbox",
        "required": false,
        "title": "Has Title",
        "type": "boolean",
      },
    },
    "title": "Feature Block",
  },
]
`;

exports[`load block from single file 1`] = `
Object {
  "group": "general",
  "icon": "extension",
  "name": "TextBlock",
  "path": "text-block",
  "props": Object {
    "fontSize": Object {
      "default": 12,
      "editor": "number",
      "options": Object {},
      "required": false,
      "title": "Font Size",
      "type": "number",
    },
    "isDark": Object {
      "default": false,
      "editor": "checkbox",
      "options": Object {},
      "required": false,
      "title": "Is Dark",
      "type": "boolean",
    },
    "items": Object {
      "default": Array [],
      "editor": "textarea",
      "options": Object {},
      "required": false,
      "title": "Items",
      "type": "array",
    },
    "text": Object {
      "default": "default text",
      "editor": "text",
      "options": Object {},
      "required": false,
      "title": "Text",
      "type": "string",
    },
  },
  "title": "Text",
}
`;

exports[`load block from single js file 1`] = `
Object {
  "group": "general",
  "name": "FeatureBlock",
  "path": "feature-block",
  "props": Object {
    "color": Object {
      "default": "",
      "editor": "text",
      "required": true,
      "title": "Color",
      "type": "string",
    },
    "hasTitle": Object {
      "default": false,
      "editor": "checkbox",
      "required": false,
      "title": "Has Title",
      "type": "boolean",
    },
  },
  "title": "Feature Block",
}
`;

exports[`load blocks from directory 1`] = `
Array [
  Object {
    "group": "general",
    "name": "FeatureBlock",
    "path": "feature-block",
    "props": Object {
      "color": Object {
        "default": "",
        "editor": "text",
        "required": true,
        "title": "Color",
        "type": "string",
      },
      "hasTitle": Object {
        "default": false,
        "editor": "checkbox",
        "required": false,
        "title": "Has Title",
        "type": "boolean",
      },
    },
    "title": "Feature Block",
  },
  Object {
    "group": "images",
    "icon": "image",
    "name": "ImageBlock",
    "path": "image-block",
    "props": Object {},
    "title": "Image",
  },
  Object {
    "group": "general",
    "icon": "image",
    "name": "TeaserBlock",
    "path": "teaser-block",
    "props": Object {
      "caption": Object {
        "default": "default text",
        "editor": "text",
        "options": Object {},
        "required": false,
        "title": "Caption",
        "type": "string",
      },
      "variant": Object {
        "default": "red",
        "editor": "radio-group",
        "options": Object {
          "variants": Array [
            "red",
            "blue",
            Object {
              "label": "Green variant",
              "value": "green",
            },
          ],
        },
        "required": false,
        "title": "Variants",
        "type": "string",
      },
    },
    "title": "Teaser block",
  },
  Object {
    "group": "general",
    "icon": "extension",
    "name": "TextBlock",
    "path": "text-block",
    "props": Object {
      "fontSize": Object {
        "default": 12,
        "editor": "number",
        "options": Object {},
        "required": false,
        "title": "Font Size",
        "type": "number",
      },
      "isDark": Object {
        "default": false,
        "editor": "checkbox",
        "options": Object {},
        "required": false,
        "title": "Is Dark",
        "type": "boolean",
      },
      "items": Object {
        "default": Array [],
        "editor": "textarea",
        "options": Object {},
        "required": false,
        "title": "Items",
        "type": "array",
      },
      "text": Object {
        "default": "default text",
        "editor": "text",
        "options": Object {},
        "required": false,
        "title": "Text",
        "type": "string",
      },
    },
    "title": "Text",
  },
]
`;
